src/lib.rs

```diff
@@ -1,168 +1,236 @@
+//! # nix-config-parser
+//!
+//! A simple parser for the Nix configuration file format.
 use std::{
     collections::HashMap,
     path::{Path, PathBuf},
 };
 
 use thiserror::Error;
 
+/// A newtype wrapper around a [`String`] that represents a `nix.conf` setting
+/// key.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct NixConfigKey(pub String);
+
+impl std::fmt::Display for NixConfigKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.0)
+    }
+}
+
+impl From<&str> for NixConfigKey {
+    fn from(value: &str) -> Self {
+        Self(value.to_string())
+    }
+}
+impl From<String> for NixConfigKey {
+    fn from(value: String) -> Self {
+        Self(value)
+    }
+}
+
+/// A newtype wrapper around a [`String`] that represents a `nix.conf` setting
+/// value.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct NixConfigValue(pub String);
+
+impl std::fmt::Display for NixConfigValue {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.0)
+    }
+}
+
+impl From<&str> for NixConfigValue {
+    fn from(value: &str) -> Self {
+        Self(value.to_string())
+    }
+}
+impl From<String> for NixConfigValue {
+    fn from(value: String) -> Self {
+        Self(value)
+    }
+}
+
 /// A newtype wrapper around a [`HashMap`], where the key is the name of the Nix
 /// setting, and the value is the value of that setting. If the setting accepts
 /// a list of values, the value will be space delimited.
 #[derive(Clone, Eq, PartialEq, Debug, Default)]
-pub struct NixConfig(pub HashMap<String, String>);
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(transparent))]
```

Comment:
I think this is the thing I'm most meh on... If we add new fields, we wouldn't be able to do this anymore (without implementing the `serde` stuff by hand. I could drop this attr, and the receipt over in `nix-installer` would then look like `nix_config: { settings: {  experimental-features = flakes, cores = 42 } }`, etc...